packages/runtime-common/loader.ts

```diff
@@ -435,11 +435,20 @@ export class Loader {
     urlOrRequest: string | URL | Request,
     init?: RequestInit,
   ): Promise<Response> {
+    let mergedHeaders: HeadersInit =
+      urlOrRequest instanceof Request
+        ? headersFromRequest(urlOrRequest)
+        : init?.headers ?? {};
     try {
       for (let handler of this.urlHandlers) {
         let request = this.asRequest(urlOrRequest, init);
+        mergedHeaders = { ...mergedHeaders, ...headersFromRequest(request) };
+        setHeaders(request, mergedHeaders);
```

Comment:
It’s inside the loop. Each handler has the ability to contribute header, so we carry that forward. The challenge is that there is not a stable request object being mutated with each handler (which is how middleware normally works). I tried changing that to a single request object that passes thru all the handlers. But there are some situations where we read the body stream and in doing so you prevent the request from actually working (you “spent” the request). So I tried as light a touch as possible here gathering up all the headers after each request regardless if it was a different request instance that what we started with and merging requests into each request before the handler runs and gathering up any new headers after it runs.

Reply:
Absolutely positively yes it is needed!!!!!

in this PR we actually set a special header to tell the system the request need to use a WIP index instead of a production index. That header should be in combination with another headers url handlers provide. The expectation here is that this is middleware.

Reply:
Oh I see it now `X-Boxel-Use-WIP-Index`